lib/rucio/daemons/reaper/reaper2.py

```diff
@@ -0,0 +1,359 @@
+                else:
+                    if needed_free_space:
+                        dict_rses[(rse['rse'], rse['id'])] = [needed_free_space, max_being_deleted_files]
+                        tot_needed_free_space += needed_free_space
+                    else:
+                        logging.debug('%s Nothing to delete on %s', prepend_str, rse['rse'])
+
+            # Ordering the RSEs based on the needed free space
+            sorted_dict_rses = OrderedDict(sorted(dict_rses.items(), key=itemgetter(1), reverse=True))
+            logging.debug('%s List of RSEs to process ordered by needed space desc : %s', prepend_str, str(sorted_dict_rses))
+
+            # Get the mapping between the RSE and the hostname used for deletion. The dictionary has RSE as key and (hostanme, rse_info) as value
+            rses_hostname_mapping = get_rses_to_hostname_mapping()
+            # logging.debug('%s Mapping RSEs to hostnames used for deletion : %s', prepend_str, str(rses_hostname_mapping))
+
+            # Loop over the RSEs. rse_key = (rse, rse_id)
+            for rse_key in sorted_dict_rses:
+                rse_name, rse_id = rse_key
+                max_workers = ceil(sorted_dict_rses[rse_key][0] / tot_needed_free_space * heart_beat['nr_threads'] * 10)
+                logging.debug('%s Working on %s. Percentage of the total space needed %.2f. Will use at most %i workers', prepend_str, rse_name, sorted_dict_rses[rse_key][0] / tot_needed_free_space * 100, max_workers)
+                payload_cnt = list_payload_counts(executable, older_than=600, hash_executable=None, session=None)
+                logging.debug('%s Payload count : %s', prepend_str, str(payload_cnt))
+                rse_hostname, rse_info = rses_hostname_mapping[rse_name]
+                rse_hostname_key = '%s,%s' % (rse_name, rse_hostname)
+
+                tot_threads_for_hostname = 0
+                for key in payload_cnt:
+                    if key and key.find(',') > -1 and key.split(',')[1] == rse_hostname:
+                        tot_threads_for_hostname += payload_cnt[key]
+
+                payload_threads = get_payload_partition(executable, hostname, pid, hb_thread, payload=rse_hostname_key, older_than=600, hash_executable=None, session=None)
+
+                if rse_hostname_key in payload_cnt and (tot_threads_for_hostname >= max_deletion_thread or payload_threads['nr_threads'] >= max_workers):
+                    logging.debug('%s Too many deletion threads for %s on RSE %s. Back off', prepend_str, rse_hostname, rse_name)
+                    # Might need to reschedule a try on this RSE later in the same cycle
+                    continue
+                del_start_time = time.time()
+                logging.info('%s Starting new worker on RSE %s', prepend_str, rse_name)
+                live(executable, hostname, pid, hb_thread, older_than=600, hash_executable=None, payload=rse_hostname_key, session=None)
```

Comment:
We should probably right away use rse_ids in the payload, or at least rsename,rse_id tuples